File: src_qt/rules/DeleteNode.cpp
#include "DeleteNode.h"
#include "../Brain.h"

namespace CnotiMind
{

	DeleteNode::DeleteNode(const QString& key, const QString& value, EventPosition position,
						   MemoryType memory, Brain* brain, QObject* parent ) :
		RuleNode(brain, parent),
		_key( key ),
		_value( value ),
		_memory( memory ),
		_position( position )
	{
	}

	void DeleteNode::exec()
	{
		if(_value.isEmpty())
		{
			_brain->deleteEvent( _key, _position, _memory );
		}
		else
		{
			_brain->deleteEvent( _key, _value, _position, _memory );
		}
	}

	void DeleteNode::exec( QHash<QString, QString>& variables )
	{

		if(_value.isEmpty())
		{
			_brain->deleteEvent( _key, _position, _memory );
		}
		else
		{
			// Test if there are variables in the value
			QString value = _value;
			tagsToValue( value, variables );

			_brain->deleteEvent( _key, value, _position, _memory );
		}
	}

	QString DeleteNode::info( int depth ) const
	{
		return space(depth) + "Delete Node " + _key + ": " + _value + " (" + (_memory == LongTermMemory ? "LTM" : "WM" ) + ")";
	}


	DeleteNode *DeleteNode::fromXML( const QString &qName, const QXmlAttributes &atts, Brain* brain, QObject* parent )
	{
		if( qName.compare( "", Qt::CaseInsensitive ) == 0 )
		{
			QString name = atts.value( "name" );
			QString value = atts.value( "value" );
			EventPosition position = translateEventPosition( atts.value( "position" ) ) ;
			MemoryType memory = translateMemoryType( atts.value( "memory" ) );

			return new DeleteNode( name, value, position, memory, brain, parent );

		}
	}
}

File: src_ios/CnotiMind/rules/DeleteNode.m
//
//  DeleteNode.m
//  SingleEmotion
//
//  Created by Gonçalo Rodrigues on 8/31/11.
//  Copyright 2011 Universidade de Coimbra. All rights reserved.
//

#import "DeleteNode.h"
#import "Brain.h"

@implementation DeleteNode

- (id)init
{
    self = [super init];
    if (self) {
        // Initialization code here.
    }
    
    return self;
}


- (id) initWithKeyAndValueAndPositionAndMemoryAndBrainAndParent:(NSString*)aKey
                                                          value:(NSString*)aValue
                                                       position:(enum DeletePosition)aPosition
                                                         memory:(enum MemoryType)aMemory 
                                                          brain:(Brain*)aBrain 
                                                         parent:(id)aParent
{
    if (self = [super initWithBrainAndParent:aBrain parent:aParent]) {
        
        _key = (aKey == nil) ? [[NSString alloc] initWithString:@""] : [[NSString alloc] initWithString:aKey];
        _value = (aValue == nil) ? [[NSString alloc] initWithString:@""] : [[NSString alloc] initWithString:aValue];
        _memory = aMemory;
        _position = aPosition;
    }
    return self;
}


- (void) exec
{
    if([_value length]==0)
    {
        [_brain deleteEvent:_key position:_position memory:_memory];
    }
    else
    {
        [_brain deleteEvent:_key value:_value position:_position memory:_memory];
    }
}


- (void) exec:(NSMutableDictionary*)aVariables
{
    if([_value length]==0)
    {
        [_brain deleteEvent:_key position:_position memory:_memory];
    }
    else
    {
        // Test if there are variables in the value
        NSString* value = _value;
        [self tagsToValue:&value variables:&aVariables];
        
        [_brain deleteEvent:_key value:value position:_position memory:_memory];
    }
}


- (NSString*) info:(int)aDepth
{
    NSString* info = [NSString stringWithFormat:@""];
    NSString* space = [self space:aDepth];
    
    info = [info stringByAppendingFormat:@"%@ DeleteNode key:%@ value:%@ (%@)", space, _key, _value, _memory == LongTermMemory ? @"LTM" : @"WM"];
    info = [info stringByAppendingFormat:@"%@", [super info:aDepth]];
    
    return info;
}


- (void) dealloc
{
    [super dealloc];
}


@end
